src/include/btree.i

```diff
@@ -35,6 +35,31 @@ __wt_page_is_modified(WT_PAGE *page)
 #define	WT_ALLOC_OVERHEAD	32U
 
 /*
+ * Track a field in the cache. Use atomic CAS so that we can reliably avoid
+ * decrementing the cache below zero - since we use an unsigned number.
+ * Track if we would go below zero in a diagnostic build - something has gone
+ * wrong.
+ */
+#ifdef	HAVE_DIAGNOSTIC
+#define WT_CACHE_DECR(session, f, sz) do {				\
+	uint64_t __val = f;						\
+	uint64_t __sz = WT_MIN(__val, sz);				\
+	if (__sz < sz)							\
+		__wt_errx(session, "%s underflow: decrementing %"	\
+		    WT_SIZET_FMT, #f, sz);				\
+	while (!WT_ATOMIC_CAS8(f, __val, __val - __sz))			\
+		__val = f, __sz = WT_MIN(__val, __sz);			\
+} while (0)
+#else
+#define WT_CACHE_DECR(session, f, sz) do {				\
+	uint64_t __val = f;						\
+	uint64_t __sz = WT_MIN(__val, sz);				\
+	while (!WT_ATOMIC_CAS8(f, __val, __val - __sz))			\
+		__val = f, __sz = WT_MIN(__val, __sz);			\
+} while (0)
+#endif
+
+/*
```

Comment:
On second thought... I'm worried we're going to see a lot of races here, this is a reasonably heavily-contested area.

Maybe using the "back it out if it underflows" change I put in 23eac61 would be better, since it's unlikely to penalize most applications -- the down-side is that we do underflow for a short period of time, and the test isn't as good. Thoughts?